Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -145,8 +145,217 @@ public struct HTTPResponseHead: Equatable {
     }
 }
 
-fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowerCasedName: [(originalCaseName, value)]
+/// The Index for a header name or value that points into the underlying `ByteBuffer`.
+typealias HTTPHeaderIndex = (start: Int, length: Int)
+
+/// Internal storage used for `HTTPHeaders`.
+private final class HTTPHeadersByteStorage : CustomStringConvertible {
```

Comment:
@weissi @Lukasa We could also just directly merge this with `HTTPHeaders` I think as `ByteBuffer` and `Array` will already ensure COW semantics.